include/ur_api.h

```diff
@@ -4111,6 +4112,48 @@ urGetLastResult(
 #if !defined(__GNUC__)
 #pragma region program
 #endif
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Program metadata property type.
+typedef enum ur_program_metadata_type_t {
+    UR_PROGRAM_METADATA_TYPE_UNKNOWN = 0,    ///< type is unknown.
+    UR_PROGRAM_METADATA_TYPE_UINT32 = 1,     ///< type is a 32-bit integer.
+    UR_PROGRAM_METADATA_TYPE_UINT64 = 2,     ///< type is a 32-bit integer.
+    UR_PROGRAM_METADATA_TYPE_BYTE_ARRAY = 3, ///< type is a byte array.
+    UR_PROGRAM_METADATA_TYPE_STRING = 4,     ///< type is a null-terminated string.
+    UR_PROGRAM_METADATA_TYPE_FORCE_UINT32 = 0x7fffffff
+
+} ur_program_metadata_type_t;
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Program metadata value union.
+typedef union ur_program_metadata_value_t {
+    char *pString; ///< [in] pointer to null-terminated string data.
+    void *pData;   ///< [in] pointer to binary data.
+    uint64_t data; ///< [in] inline storage for the value.
+
+} ur_program_metadata_value_t;
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Program metadata property.
+typedef struct ur_program_metadata_t {
+    char *pName;                       ///< [in] null-terminated metadata name.
+    ur_program_metadata_type_t type;   ///< [in] the type of metadata value.
+    size_t size;                       ///< [in] size in bytes of the data pointed to by value.pData, or 0 when
+                                       ///< value size is less than 64-bits and is stored directly in value.data.
+    ur_program_metadata_value_t value; ///< [in] the metadata value storage.
+
+} ur_program_metadata_t;
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Program creation properties.
+typedef struct ur_program_properties_t {
+    ur_structure_type_t stype;               ///< [in] type of this structure
+    void *pNext;                             ///< [in,out][optional] pointer to extension-specific structure
+    uint32_t count;                          ///< [in] the number of entries in pMetadatas.
```

Comment:
should we clarify that if count is non-zero, then pMetaDatas should be non-null? kind of obvious but maybe need to add it?

Reply:
Sounds like a good idea, Ideally there would be error codes for this also. I've not figured out how to do that with the generator yet.

Reply:
Updated the documentation and also added error codes to `urProgramCreate[WithBinary]`.